packages/starlight/integrations/virtual-user-config.ts

```diff
@@ -82,3 +88,37 @@ export function vitePluginStarlightUserConfig(
 		},
 	};
 }
+
+function generateGitInfoModule({
+	command,
+	srcDir,
+}: {
+	command: 'dev' | 'build' | 'preview';
+	srcDir: string;
+}) {
+	const docsPath = resolve(srcDir, 'content/docs');
+
+	if (command !== 'build') {
+		const onDemandGitInfoModule = new URL('../utils/git.ts', import.meta.url);
+
+		// In dev mode expose the function directly so git is executed only when needed.
+		return `import {getNewestCommitDate as getInternal} from '${onDemandGitInfoModule}';
+import { resolve } from 'node:path';
+
+export const getNewestCommitDate = (file) => getInternal(
+	resolve(${JSON.stringify(docsPath)}, file)
+);
+`;
+	}
+
+	const trackedDocsFiles = getAllNewestCommitDate(docsPath);
+
+	return `
+const trackedDocsFiles = new Map(${JSON.stringify(trackedDocsFiles)});
+
+export function getNewestCommitDate(file) {
+	const timestamp = trackedDocsFiles.get(file);
+	if (!timestamp) throw new Error(\`Failed to retrieve the git history for file "\${file}"\`);
+	return new Date(timestamp);
+}`;
+}
```

Comment:
In general all the other modules in this file aim to keep the generative stuff as minimal and clean as possible. Would it be possible to refactor this to move as much as possible to imported code? That way we get the benefit of typechecking, editor tooling etc. and avoid too much JS written in template strings.

i.e. a quick naïve take, make a file a bit like the following (or these could also be inlined into `utils/git` itself):

```ts
import { getNewestCommitDate, getAllNewestCommitDate } from '../utils/git';

export function getDevGitUtil(docsPath: string) {
	return (file: string) => getNewestCommitDate(resolve(docsPath, file));
}
export function getBuildGitUtil(docsPath: string) {
	const trackedDocsFiles = new Map(getAllNewestCommitDate(docsPath));
	return (file: string) => {
		const timestamp = trackedDocsFiles.get(file);
		if (!timestamp) throw new Error(`Failed to retrieve the git history for file "${file}"`);
		return new Date(timestamp);
	};
}
```

and then the virtual module does the bare minimum:

```ts
const resolvedDocsPath = resolveId(srcDir, 'content/docs');
const gitUtilName = command !== 'build' ? 'getDevGitUtil' : 'getBuildGitUtil';
const modules = {
	// ...
	'virtual:starlight/git-info': `import { ${gitUtilName} as getGitUtil } from "...";
export const getNewestCommitDate = getGitUtil(${resolvedDocsPath});`
	// ...
};
```

Reply:
The point is exactly to the git module during build and inline the result, so there is no dependency on git or Node APIs during runtime.

Calling `getAllNewestCommitDate` from the virtual module (even if indirectly) inverts the whole thing, it would compute the git information every time the server initializes.

Reply:
OK! Then can we run with the general principle of simplifying? i.e. still compute those for the build version, but then the above example looks something like:
```ts
const argument = command !== 'build' ? resolveId(srcDir, 'content/docs') : getAllNewestCommitDate(...);
const gitUtilName = command !== 'build' ? 'getDevGitUtil' : 'getBuildGitUtil';
const modules = {
	// ...
	'virtual:starlight/git-info': `import { ${gitUtilName} as getGitUtil } from "...";
export const getNewestCommitDate = getGitUtil(${argument});`
	// ...
};
```

As I mentioned, my code was a “naïve take,” but the principle is to move as much as possible out of a template literal and into regular code.